src/mercury_engine_data_structures/formats/rom3ds.py

```diff
@@ -0,0 +1,342 @@
+)
+
+RomFsDirectoryEntry = Struct(
+    "parent_offset" / Int32ul,
+    "sibling_offset" / Int32ul,
+    "child_offset" / Int32ul,
+    "file_offset" / Int32ul,
+    "hash_sibling_offset" / Int32ul,
+    "name" / PascalString(Int32ul, "utf_16_le"),
+)
+
+RomFsFileEntry = Struct(
+    "parent_offset" / Int32ul,
+    "sibling_offset" / Int32ul,
+    "data_offset" / Int64ul,
+    "data_size" / Int64ul,
+    "hash_sibling_offset" / Int32ul,
+    "name" / PascalString(Int32ul, "utf_16_le"),
+)
+
+
+class DirectoryEntry:
+    def __init__(self, parent_path: str, entry: construct.Container):
+        self._entry = entry
+        if parent_path == "":
+            self.complete_path = self._entry.name
+        else:
+            self.complete_path = parent_path + "/" + self._entry.name
+
+
+class FileEntry:
+    def __init__(self, parent_path: str, entry: construct.Container):
+        self._entry = entry
+        if parent_path == "":
+            self.complete_path = self._entry.name
+        else:
+            self.complete_path = parent_path + "/" + self._entry.name
+        self.data_size = entry.data_size
+        self.data_offset = entry.data_offset
+
+
+class Rom3DS:
+    def __init__(self, file_name: str, file_stream: BufferedReader):
+        self.file_name_to_entry = {}
+        self.file_stream = file_stream
+        try:
+            if file_name.lower().endswith("cia"):
+                self.raw = AddCia.parse_stream(file_stream)
+            elif file_name.lower().endswith("3ds"):
+                self.raw = Add3ds.parse_stream(file_stream)
+            elif file_name.lower().endswith("app") or file_name.lower().endswith("cxi"):
+                self.raw = AddCxi.parse_stream(file_stream)
+            else:
+                raise ValueError('Input does not end with ".cia", ".3ds", ".app" or ".cxi"')
+        # encrypted files should throw a ConstError because the ".code" string is encrypted and parsing fails
+        except construct.core.ConstError:
+            raise ValueError("Rom file could not be parsed. Make sure that you use a decrypted supported file format.")
+        self.read_rom_fs()
+
+    def get_title_id(self) -> str:
+        title_id = self.raw.rom_struct.ncch.program_id[::-1]
+        return title_id.hex().upper()
+
+    def is_pal(self) -> bool:
+        # there is probably a more general method but region info is embedded in the icon. out of scope for now
+        return self.get_title_id() == "00040000001BFB00"
+
+    def _is_code_binary_compressed(self) -> bool:
+        # first 8 bytes are the name, followed by 5 reserved bytes, followed by a flag byte where the first bit
+        # tells if compressed
+        is_compressed = self.raw.rom_struct.ncchexheader[13] & 0x01
+        return is_compressed
+
+    def get_file_binary(self, path: str) -> bytes:
+        entry = self.file_name_to_entry.get(path, None)
+        if entry is None:
+            raise ValueError(f"Could not find file with path {path}")
```

Comment:
Maybe just do `self.file_name_to_entry[path]` if you want to raise on missing fields?